composer/callbacks/mlperf.py

```diff
@@ -0,0 +1,345 @@
+                raise ValueError("Batch size is required to be set for dataloader.")
+            if len(state.evaluators) > 1:
+                raise ValueError("Only one evaluator is supported for the MLPerfCallback.")
+            if not isinstance(state.train_dataloader.dataset, Sized):
+                raise ValueError("Train dataset must have __len__ property")
+            if not isinstance(state.evaluators[0].dataloader.dataloader.dataset, Sized):
+                raise ValueError("Train dataset must have __len__ property")
+
+            self._log_dict({
+                constants.SEED: state.seed,
+                constants.GLOBAL_BATCH_SIZE: state.train_dataloader.batch_size * dist.get_world_size(),
+                constants.GRADIENT_ACCUMULATION_STEPS: state.grad_accum,
+                constants.TRAIN_SAMPLES: len(state.train_dataloader.dataset),
+                constants.EVAL_SAMPLES: len(state.evaluators[0].dataloader.dataloader.dataset)
+            })
+
+        self.mllogger.event(key=constants.INIT_STOP)
+
+        dist.barrier()
+        if rank_zero():
+            self.mllogger.event(key=constants.RUN_START)
+
+    def epoch_start(self, state: State, logger: Logger) -> None:
+        if rank_zero():
+            self.mllogger.event(key=constants.EPOCH_START, metadata={'epoch_num': state.timer.epoch.value})
+            self.mllogger.event(key=constants.BLOCK_START,
+                                metadata={
+                                    'first_epoch_num': state.timer.epoch.value,
+                                    'epoch_count': 1
+                                })
+
+    def epoch_end(self, state: State, logger: Logger) -> None:
+        if rank_zero():
+            self.mllogger.event(key=constants.EPOCH_STOP, metadata={'epoch_num': state.timer.epoch.value})
+            logger.file_artifact(LogLevel.FIT, artifact_name=self.upload_name, file_path=self.filename)
+
+    def eval_start(self, state: State, logger: Logger) -> None:
+        if rank_zero():
+            self.mllogger.event(key=constants.EVAL_START, metadata={'epoch_num': state.timer.epoch.value})
+
+    def eval_end(self, state: State, logger: Logger) -> None:
+        if rank_zero():
+            accuracy = self._get_accuracy(state)
+
+            self.mllogger.event(key=constants.EVAL_STOP, metadata={'epoch_num': state.timer.epoch.value})
+            self.mllogger.event(key=constants.EVAL_ACCURACY,
+                                value=accuracy,
+                                metadata={'epoch_num': state.timer.epoch.value})
+            self.mllogger.event(key=constants.BLOCK_STOP, metadata={'first_epoch_num': state.timer.epoch.value})
+
+            if accuracy > self.target and not self.success:
+                self.mllogger.event(key=constants.RUN_STOP, metadata={"status": "success"})
+                self.mllogger.logger.removeHandler(self._file_handler)
+                self.success = True  # only log once
+
```

Comment:
Can the callback implement `def close(self, state, logger)` and ensure that any opened files are closed (such as the `self._file_handler`)? The trainer will invoke this when shutting down to ensure that open file handlers are not being left behind. It is OK to close the file earlier; but we should also close it here (if the file is still open) in case if there was some exception somewhere.